service/src/main/kotlin/xyz/lebkuchenfm/external/storage/dropbox/DropboxFileStorage.kt

```diff
@@ -0,0 +1,132 @@
+import io.ktor.client.plugins.auth.Auth
+import io.ktor.client.plugins.auth.providers.BearerTokens
+import io.ktor.client.plugins.auth.providers.bearer
+import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
+import io.ktor.client.request.accept
+import io.ktor.client.request.forms.FormDataContent
+import io.ktor.client.request.headers
+import io.ktor.client.request.post
+import io.ktor.client.request.setBody
+import io.ktor.http.ContentType
+import io.ktor.http.URLBuilder
+import io.ktor.http.contentType
+import io.ktor.http.parameters
+import io.ktor.serialization.kotlinx.json.json
+import io.ktor.server.config.ApplicationConfig
+import kotlinx.serialization.json.Json
+import xyz.lebkuchenfm.external.storage.dropbox.models.ShareFile
+import xyz.lebkuchenfm.external.storage.dropbox.models.ShareFileResponse
+import xyz.lebkuchenfm.external.storage.dropbox.models.ShareFileSettings
+import xyz.lebkuchenfm.external.storage.dropbox.models.TokenInfo
+import xyz.lebkuchenfm.external.storage.dropbox.models.UploadFileArgs
+import xyz.lebkuchenfm.external.storage.dropbox.models.UploadFileResponse
+
+class DropboxFileStorage(config: ApplicationConfig) {
+    private val client = prepareClient()
+    private val bearerTokenStorage = mutableListOf<BearerTokens>()
+    private val refreshToken = config.property(DROPBOX_REFRESH_TOKEN_PROPERTY_PATH).getString()
+    private val appKey = config.property(DROPBOX_APP_KEY_PROPERTY_PATH).getString()
+    private val appSecret = config.property(DROPBOX_APP_SECRET_PROPERTY_PATH).getString()
+
+    /**
+     * [path] should contain destination folder, file name and its extension
+     * [bytes] contains file content data
+     * @return url on which raw file may be obtained
+     */
+    suspend fun uploadFile(
+        path: String,
+        bytes: ByteArray,
+    ): String {
+        if (bearerTokenStorage.isEmpty()) {
+            getInitialTokens()
+        }
+
+        val args = UploadFileArgs(path, "add", false, false)
+        val argsString = Json.encodeToString(UploadFileArgs.serializer(), args)
+
+        val uploadFile: UploadFileResponse =
+            client.post("https://content.dropboxapi.com/2/files/upload") {
+                setBody(bytes)
+                contentType(ContentType.Application.OctetStream)
+                accept(ContentType.Application.Json)
+                headers {
+                    append("Dropbox-API-Arg", argsString)
+                }
+            }.body()
+
+        val sharedFile: ShareFileResponse =
+            client.post("https://api.dropboxapi.com/2/sharing/create_shared_link_with_settings") {
+                setBody(ShareFile(uploadFile.pathDisplay, ShareFileSettings("viewer", "public", true)))
+                contentType(ContentType.Application.Json)
+            }.body()
+
+        val resourceUrl =
+            URLBuilder(sharedFile.url).apply {
+                this.host = "dl.dropboxusercontent.com"
+                this.parameters.remove("dl")
```

Comment:
You can remove `this.` inside `apply`. I think it's clearer, but I will leave it up to you.